internal/cache/index.go

```diff
@@ -0,0 +1,336 @@
+	// Value is the value of the Object stored in the Cache
+	// It is used by Caches but not by the Index
+	Value []byte `msg:"value,omitempty"`
+}
+
+// ToBytes returns a serialized byte slice representing the Object
+func (o *Object) ToBytes() []byte {
+	bytes, _ := o.MarshalMsg(nil)
+	return bytes
+}
+
+// ObjectFromBytes returns a deserialized Cache Object from a seralized byte slice
+func ObjectFromBytes(data []byte) (*Object, error) {
+	o := &Object{}
+	_, err := o.UnmarshalMsg(data)
+	return o, err
+}
+
+// NewIndex returns a new Index based on the provided inputs
+func NewIndex(cacheName, cacheType string, indexData []byte, config config.CacheIndexConfig, bulkRemoveFunc func([]string, bool), flushFunc func(cacheKey string, data []byte)) *Index {
+	i := &Index{}
+
+	if len(indexData) > 0 {
+		i.UnmarshalMsg(indexData)
+	} else {
+		i.Objects = make(map[string]*Object)
+	}
+
+	i.name = cacheName
+	i.flushInterval = time.Duration(config.FlushIntervalSecs) * time.Second
+	i.flushFunc = flushFunc
+	i.reapInterval = time.Duration(config.ReapIntervalSecs) * time.Second
+	i.bulkRemoveFunc = bulkRemoveFunc
+	i.config = config
+
+	if i.flushInterval > 0 && flushFunc != nil {
+		go i.flusher()
+	} else {
+		log.Warn("cache index flusher did not start", log.Pairs{"cacheName": i.name, "flushInterval": i.flushInterval})
+	}
+
+	if i.reapInterval > 0 {
+		go i.reaper()
+	} else {
+		log.Warn("cache reaper did not start", log.Pairs{"cacheName": i.name, "reapInterval": i.reapInterval})
+	}
+
+	return i
+}
+
+// UpdateObjectAccessTime updates the LastAccess for the object with the provided key
+func (idx *Index) UpdateObjectAccessTime(key string) {
+	indexLock.Lock()
+	defer indexLock.Unlock()
+	if _, ok := idx.Objects[key]; ok {
+		idx.Objects[key].LastAccess = time.Now()
+	}
+}
+
+// UpdateObject writes or updates the Index Metadata for the provided Object
+func (idx *Index) UpdateObject(obj Object) {
+
+	key := obj.Key
+	if key == "" {
+		return
+	}
+
+	indexLock.Lock()
+	defer indexLock.Unlock()
+
+	obj.Size = int64(len(obj.Value))
+	obj.Value = nil
+	obj.LastAccess = time.Now()
+	obj.LastWrite = obj.LastAccess
+
+	if o, ok := idx.Objects[key]; ok {
+		idx.CacheSize += o.Size - idx.Objects[key].Size
+	} else {
+		idx.CacheSize += obj.Size
+		idx.ObjectCount++
+	}
+
+	ObserveCacheSizeChange(idx.name, idx.cacheType, idx.CacheSize, idx.ObjectCount, idx.config.MaxSizeObjects, idx.config.MaxSizeBytes)
+
+	idx.Objects[key] = &obj
+}
+
+// RemoveObject removes an Object's Metadata from the Index
+func (idx *Index) RemoveObject(key string, noLock bool) {
+
+	if !noLock {
+		indexLock.Lock()
+		defer indexLock.Unlock()
+	}
+	if o, ok := idx.Objects[key]; ok {
+		idx.CacheSize -= o.Size
+		idx.ObjectCount--
+
+		ObserveCacheOperation(idx.name, idx.cacheType, "del", "none", float64(o.Size))
+
+		delete(idx.Objects, key)
+		ObserveCacheSizeChange(idx.name, idx.cacheType, idx.CacheSize, idx.ObjectCount, idx.config.MaxSizeObjects, idx.config.MaxSizeBytes)
```

Comment:
Do you ever expect this to change after a cache has been initialized? You could move this to another metric to avoid setting a static value again and again. Then again I don't know how Prometheus storage works, it could very well be the use a common copy.

Reply:
Anyone can adjust the max cache size in their config and bounce Trickster, so we need to continually monitor it in order to properly report cache utilization / capacity.

Reply:
The alternative I was think of it setting it in its own separate metric to avoid including this information on every single unique label set for this metric. Just set it once and you are done.

Reply:
I got you. The maxes are already their own metric (the observe func just aggregates the calls). I've moved the two max metrics out of observe and over to the NewIndex(), so they'll only be set once. thanks for the greaet catch!